File: Kata.GameOfLife.JavaScript/src/game.js
'use strict';

var Cell = function (x, y) {
  this.x = x;
  this.y = y;
};

Cell.prototype.equals = function (otherCell) {
  return this.x === otherCell.x && this.y === otherCell.y;
};


var Grid = function () {
  var aliveCells = [].slice.call(arguments);

  var isAlive = function (cell) {
    return aliveCells.some(function (aliveCell) {
      return aliveCell.equals(cell);
    });
  };

  var getNeighborsOf = function (cell, filter) {
    var neighbors = [];

    for (var x = -1; x < 2; x++) {
      for (var y = -1; y < 2; y++) {
        var neighbor = new Cell(cell.x + x, cell.y + y);
        if (!neighbor.equals(cell)) {
          neighbors.push(neighbor);
        }
      }
    }

    if (filter) {
      neighbors = neighbors.filter(filter);
    }
    return neighbors;
  };

  var getAliveNeighborsOf = function (cell) {
    return getNeighborsOf(cell, function (neighbor) {
      return isAlive(neighbor);
    });
  };

  var getDeadNeighborsOf = function (cell) {
    return getNeighborsOf(cell, function (neighbor) {
      return !isAlive(neighbor);
    });
  };

  var getAliveCandidates = function () {
    return aliveCells.filter(function (aliveCell) {
      var count = getAliveNeighborsOf(aliveCell).length;
      return (count === 2 || count === 3);
    });
  };

  var getReviveCandidates = function () {
    var reviveCandidates = [];

    aliveCells.forEach(function (aliveCell) {
      var deadNeighbors = getDeadNeighborsOf(aliveCell);

      deadNeighbors.forEach(function (neighbor) {
        if (getAliveNeighborsOf(neighbor).length === 3) {
          reviveCandidates.push(neighbor);
        }
      });
    });
    return reviveCandidates;
  };

  var newGeneration = function () {
    aliveCells = getAliveCandidates().concat(getReviveCandidates());
  };

  return {
    isAlive: isAlive,
    newGeneration: newGeneration
  };
};

File: Kata.GameOfLife.Python/game.py
from cell import Cell

class Game:
  def __init__(self, *args):
    self.aliveCells = []

    if (isinstance(args[0], list)):
      self.aliveCells = args[0]
    else:
      for aliveCell in args:
        self.aliveCells.append(aliveCell)

  def isAlive(self, cell):
    return any(aliveCell.equals(cell) for aliveCell in self.aliveCells)

  def getNeighborsOf(self, cell):
    neighbors = []

    for x in range(-1, 2):
      for y in range(-1, 2):
        neighbor = Cell(cell.x + x, cell.y + y)
        if (cell.equals(neighbor) == False):
          neighbors.append(neighbor)

    return neighbors

  def getAliveNeighborsOf(self, cell):
    neighbors = self.getNeighborsOf(cell)
    return [neighbor for neighbor in neighbors if self.isAlive(neighbor)]

  def getCountOfAliveNeighborsOf(self, cell):
    return len(self.getAliveNeighborsOf(cell))

  def getAliveCandidates(self):
    aliveCandidates = []

    for aliveCell in self.aliveCells:
      count = self.getCountOfAliveNeighborsOf(aliveCell)
      if count == 2 or count == 3:
        aliveCandidates.append(aliveCell)

    return aliveCandidates

  def getDeadNeighborsOf(self, cell):
    neighbors = self.getNeighborsOf(cell)
    return [neighbor for neighbor in neighbors if self.isAlive(neighbor) == False]

  def getReviveCandidates(self):
    reviveCandidates = []

    for aliveCell in self.aliveCells:
      deadNeighbors = self.getDeadNeighborsOf(aliveCell)
      for deadNeighbor in deadNeighbors:
        if (self.getCountOfAliveNeighborsOf(deadNeighbor) == 3):
          reviveCandidates.append(deadNeighbor)

    return reviveCandidates

  def newGeneration(self):
    aliveCandidates = self.getAliveCandidates()
    reviveCandidates = self.getReviveCandidates()
    return Game(aliveCandidates + reviveCandidates)
